listings/ch01-applications/crowd_fund/src/contracts.cairo

```diff
@@ -0,0 +1,140 @@
+    fn pledge(ref self: TContractState, id: u256, amount: u256);
+    fn unpledge(ref self: TContractState, id: u256, amount: u256);
+}
+
+#[starknet::contract]
+mod CrowdFund {
+    use openzeppelin::token::erc20::interface::{IERC20Dispatcher, IERC20DispatcherTrait};
+    use starknet::{ContractAddress, get_caller_address, get_contract_address};
+    #[storage]
+    struct Storage {
+        token: IERC20Dispatcher,
+        count: u256,
+        campaigns: LegacyMap<u256, Campaign>,
+        pledged_amount: LegacyMap<(u256, ContractAddress), u256>
+    }
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        Launch: Launch,
+        Pledge: Pledge,
+        Unpledge: Unpledge,
+        Claim: Claim
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct Launch {
+        id: u256,
+        creator: ContractAddress
+    }
+    #[derive(Drop, starknet::Event)]
+    struct Pledge {
+        id: u256,
+        caller: ContractAddress,
+        amount: u256
+    }
+    #[derive(Drop, starknet::Event)]
+    struct Unpledge {
+        id: u256,
+        caller: ContractAddress,
+        amount: u256
+    }
+    #[derive(Drop, starknet::Event)]
+    struct Claim {
+        id: u256,
+        creator: ContractAddress,
+        amount: u256
+    }
+
+    #[derive(Copy, Drop, Serde, starknet::Store)]
+    struct Campaign {
+        id: u256,
+        creator: ContractAddress,
+        claimed: bool,
+        pledged: u256
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, token: ContractAddress) {
+        self.token.write(IERC20Dispatcher { contract_address: token });
+    }
+
+    #[abi(embed_v0)]
+    impl CrowdFund of super::ICrowdFund<ContractState> {
+        fn launch(ref self: ContractState) {
+            let new_campaign_id = self.count.read() + 1;
+            let caller = get_caller_address();
+            let new_campaign = Campaign {
+                id: new_campaign_id, creator: caller, claimed: false, pledged: 0
+            };
+            self.campaigns.write(new_campaign_id, new_campaign);
+            self.count.write(new_campaign_id);
+            self.emit(Launch { id: new_campaign_id, creator: caller });
+        }
+        fn claim(ref self: ContractState, id: u256) {
+            let caller = get_caller_address();
+            let this = get_contract_address();
+
+            assert(self.count.read() <= id, 'no id found');
+
+            let mut current_campaign = self.campaigns.read(id);
+            assert(current_campaign.creator == caller, 'only owner allowed');
+            assert(current_campaign.claimed == false, 'already claimed');
+
+            let token: IERC20Dispatcher = self.token.read();
+            token.transfer(caller, current_campaign.pledged);
+
+            current_campaign.claimed = true;
+            self.campaigns.write(id, current_campaign);
+
+            self.emit(Claim { id: id, creator: caller, amount: current_campaign.pledged });
```

Comment:
It's always best to follow the [Checks Effects Interactions Pattern](https://book.cairo-lang.org/ch99-03-security-considerations.html#checks-effects-interactions-pattern)

Change the `claimed` state of the current campaign only after the external `token.transfer` call.

Reply:
You mean I should verify whether the transfer was done or not by using balance_of or something else?

Reply:
The logic is good. But you should make the external calls only after all effects (i.e. `claimed = true`). This prevents reentrancy (even if it's more difficult to do in Starknet than on Ethereum).

Reply:
I can use [openzepellin](https://docs.openzeppelin.com/contracts-cairo/0.8.0/api/security#ReentrancyGuardComponent) here? using their component, I might be able to handle this?

Also I am a bit confused, you said "Change the claimed state of the current campaign only after the external token.transfer call." but now you said "But you should make the external calls only after all effects". 2nd statement is correct? right?

Reply:
Sorry my bad! I meant change the claimed state before the external call.

Reply:
You could use OZ guard but we should try to keep things simple and focus on what the example wants to show.